fix using inventory and cache plugins in a collection

s-hertel opened the pull request:
##### SUMMARY
This is now just a minimal fix for using cache plugins in collections for inventory.

Move set_options (which is reliant on having the _load_name set) outside of the CacheModule's `__init__`, and remove the janky _load_name hack.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
plugins

s-hertel committed:
Allow custom inventory plugins and cache plugins

If _load_name is not set correctly the cache plugin can't load the documentation (which is also the arg spec)

Fix the existing inventory plugin in the collections tests

Add integration tests for using a cache plugin in a collection

s-hertel commented on lib/ansible/plugins/inventory/__init__.py:
```diff
@@ -219,7 +219,8 @@ def _read_config_data(self, path):
             raise AnsibleParserError("%s is empty" % (to_native(path)))
-        elif config.get('plugin') != self.NAME:
+        # If there's a '.' in the plugin name, only fail if there are also no matching collections
+        elif config.get('plugin') != self.NAME and config.get('plugin').split('.')[-1] != self.NAME:
```
Not sure of the best way to do this. The alternative would be for NAME in the inventory module to be required to be the full collection name. @bcoca thoughts?

bcoca replied:
you should not need to do this, plugin loader should take care of collection loading itself #56194 should fix the issue of finding 'adjacent' collections for 'non ansible-playbook' , like ansible-inventory

bcoca replied:
just looked at tests, if this is also an issue with playbook, its a bug in loader, the plugins themselves should not need to be 'collection aware'

s-hertel commented:
@chrismeyersfsu if you get a chance to test this out, would love your feedback.

bcoca commented on lib/ansible/plugins/cache/__init__.py:
```diff
@@ -59,7 +59,8 @@ class BaseCacheModule(AnsiblePlugin):
-        self._load_name = self.__module__.split('.')[-1]
+        # get everything after ansible.plugins.cache
+        self._load_name = '.'.join(self.__module__.split('.')[3:])
         super(BaseCacheModule, self).__init__()
```
i would say the bug is that the plugin inside the collection is not using a correct  name, it needs to match the collection name, we should not have to make every other plugin or base class collection aware

s-hertel replied:
This was true for the inventory plugin in the collection. Thanks for noticing that. Cache is a different case since they do not set their own name. Also kind of feeling like I wish inventory plugins did not set NAME: https://github.com/ansible/ansible/pull/56469#discussion_r285124586

I modified the loader instead to set self._load_name early enough for the plugin `__init__`

s-hertel commented on test/integration/targets/collections/collections/ansible_collections/testns/content_adj/plugins/cache/custom_jsonfile.py:
```diff
@@ -52,7 +52,6 @@ class CacheModule(BaseFileCacheModule):
     """
     A caching module backed by json files.
     """
-    NAME = 'testns.content_adj.custom_jsonfile'
```
If we require cache plugins to set a load name, then they also aren't portable between collections as NAME would need to be tailored to the collection name. The more I look at this the more I like using _load_name. I think inventory plugins should also get rid of NAME for the same reason.

s-hertel commented on lib/ansible/plugins/loader.py:
```diff
@@ -588,6 +587,9 @@ def get(self, name, *args, **kwargs):
             try:
+                # A plugin may need to use its _load_name in __init__ (for example, to use self.set_options or self.get_option).
+                # Update the object before instantiating the class in case this is true
+                self._update_object(obj, name, path)
```
I left the _update_object call below for backwards compatibility in case an `__init__` has a side effect.

s-hertel replied:
An alternative to this line would be to use class_only=True where our code uses cache_loader (and any other loaders that might need _load_name in `__init__`) so _update_object is called before instantiation. I picked this because it also fixes possible third party code using the plugin loader.

s-hertel commented:
@bcoca do you mind taking another look at this when you have time?

On devel collection inventory/cache plugins can't use config. The collection inventory tests are passing on devel because the inventory plugin has been using `loader.load_from_file` and then referencing the dict to get options. This is why the test inventory plugin docs have been able to be incorrect (hostname is not a list) and the plugin had to do its own validation (rather than being able to use `required: true`). This issue also extends to using a collection cache plugin that attempts that uses config (fact caching will work if the cache plugin uses constants instead, but will get a deprecation warning) and inventory caching (which requires the plugins to use config and will not work with constants). This fixes all of that.

I'd also like to propose that we deprecate the `NAME` class attribute in inventory plugins in favor of just using the _load_name since plugins shouldn't need to be catered to the collection they are in, but I think that should be a separate PR.

s-hertel committed:
Set the attribute on the instance instead of the class

Deprecate importing custom CacheModules directly - they should use the cache_loader
